{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Исключения\n",
    "========\n",
    "\n",
    "Стоит сразу оговорить, что в современном стиле програмирования принято пользовательские исключения создавать как экземпляры классов.\n",
    "\n",
    "Однако, мы всеже рассмотрим конструкцию **try/except/else/finally** и \n",
    "\n",
    "https://docs.python.org/3/tutorial/errors.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Основные исключения**\n",
    "\n",
    "- **Exception** – то, на чем фактически строятся все остальные ошибки;\n",
    "- **AttributeError** – возникает, когда ссылка атрибута или присвоение не могут быть выполнены;\n",
    "- **IOError** – возникает в том случае, когда операция I/O (такая как оператор вывода, встроенная функция open() или метод объекта-файла) не может быть выполнена, по связанной с I/O причине: «файл не найден», или «диск заполнен», иными словами.\n",
    "- **ImportError** – возникает, когда оператор import не может найти определение модуля, или когда оператор не может найти имя файла, который должен быть импортирован;\n",
    "- **IndexError** – возникает, когда индекс последовательности находится вне допустимого диапазона;\n",
    "- **KeyError** – возникает, когда ключ сопоставления (dictionary key) не найден в наборе существующих ключей;\n",
    "- **KeyboardInterrupt** – возникает, когда пользователь нажимает клавишу прерывания(обычно Delete или Ctrl+C);\n",
    "- **NameError** – возникает, когда локальное или глобальное имя не найдено;\n",
    "- **OSError** – возникает, когда функция получает связанную с системой ошибку;\n",
    "- **SyntaxError** — возникает, когда синтаксическая ошибка встречается синтаксическим анализатором;\n",
    "- **TypeError** – возникает, когда операция или функция применяется к объекту несоответствующего типа. Связанное значение представляет собой строку, в которой приводятся подробные сведения о несоответствии типов;\n",
    "- **ValueError** – возникает, когда встроенная операция или функция получают аргумент, тип которого правильный, но неправильно значение, и ситуация не может описано более точно, как при возникновении **IndexError**;\n",
    "- **ZeroDivisionError** – возникает, когда второй аргумент операции division или modulo равен нулю;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Инструкция try/except/else**\n",
    "\n",
    "Инструкция try – это составная инструкция. Полная ее форма приводится\n",
    "ниже:\n",
    "\n",
    "    try:\n",
    "        <statements>     # Сначала выполняются эти действия\n",
    "    except <name1>:\n",
    "        <statements>     # Запускается, если в блоке try возникло исключение name1\n",
    "    except (name2, name3):\n",
    "        <statements>     # Запускается, если возникло любое из этих исключений\n",
    "    except <name4> as <data>:\n",
    "        <statements>     # Запускается в случае исключения name4\n",
    "                         # и получает экземпляр исключения\n",
    "    except:\n",
    "        <statements>     # Запускается для всех (остальных) возникших исключений\n",
    "    else:\n",
    "        <statements>     # Запускается, если в блоке try не возникло исключения"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "В этой инструкции блок под заголовком **try** представляет основное  действие\n",
    "инструкции – программный код, который следует попытаться выполнить.\n",
    "\n",
    "Предложения **except** определяют обработчики исключений, возникших в ходе\n",
    "выполнения блока **try**\n",
    "\n",
    "а предложение **else** (если присутствует) определяет обработчик для случая отсутствия исключений."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You cannot divide by zero!\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    1 / 0\n",
    "except ZeroDivisionError:\n",
    "    print(\"You cannot divide by zero!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This key is not in the dictionary!\n"
     ]
    }
   ],
   "source": [
    "my_dict = {\"a\":1, \"b\":2, \"c\":3}\n",
    " \n",
    "try:\n",
    "    value = my_dict[\"d\"]\n",
    "except IndexError:\n",
    "    print(\"This index does not exist!\")\n",
    "except KeyError:\n",
    "    print(\"This key is not in the dictionary!\")\n",
    "except:\n",
    "    print(\"Some other error occurred!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Something wrong: can only concatenate list (not \"str\") to list\n",
      "we`re still working\n"
     ]
    }
   ],
   "source": [
    "def kaboom(x, y):\n",
    "    print(x + y)       # Возбуждает исключение TypeError\n",
    "\n",
    "try:\n",
    "    kaboom([0,1,2], 'spam')\n",
    "except TypeError as err:      # Исключение перехватывается и обрабатывается здесь\n",
    "    print('Something wrong:', err)\n",
    "    \n",
    "print('we`re still working') # Программа продолжает работу независимо от того,\n",
    "                             # было ли исключение или нет"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Возможный ход развития событий**:\n",
    "\n",
    " 1. Если исключение возникнет во время выполнения инструкций в блоке **try**, интерпретатор вернется к инструкции **try** и выполнит первое предложение **except**, соответствующее возбужденному исключению. После выполнения блока **except** управление будет передано первой инструкции, находящейся за всей инструкцией **try** (при условии, что в блоке except не возникло другого исключения).\n",
    "\n",
    " 2. Если в блоке **try** возникло исключение и не было найдено ни одного соответствия среди предложений except, исключение будет передано инструкции **try**, стоящей выше в программе, или на верхний уровень процесса (что вынудит интерпретатор аварийно завершить работу программы и вывести сообщение об ошибке по умолчанию).\n",
    " \n",
    " 3. Если в процессе выполнения блока **try** не возникло исключение, интерпретатор выполнит инструкции в блоке **else** (если имеются) и затем выполнение продолжится с первой инструкции, находящейся за всей инструкцией **try**\n",
    " \n",
    " \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Оператор finally**\n",
    "\n",
    " **try/finally** специализируется на выполнении заключительных операций. Если в инструкцию **try** включено предложение **finally**, интерпретатор всегда будет выполнять этот блок инструкций при «выходе» из инструкции **try** независимо от того, произошло ли исключение во время выполнения инструкций в блоке **try**. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You cannot divide by zero!\n",
      "I need to say something!!!\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    1 / 0\n",
    "except ZeroDivisionError:\n",
    "    print(\"You cannot divide by zero!\")\n",
    "finally:\n",
    "    print(\"I need to say something!!!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Форму **try/finally** удобно использовать, когда **необходимо гарантировать выполнение некоторых действий независимо от реакции программы на исключение**. С практической точки зрения, эта форма инструкции позволяет определять завершающие действия, которые должны выполняться всегда, такие как\n",
    "закрытие файлов или закрытие соединений с сервером."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Инструкция raise**\n",
    "\n",
    "Чтобы явно возбудить исключение, можно использовать инструкцию **raise**.\n",
    "В общем виде она имеет очень простую форму записи – инструкция **raise** состоит из слова **raise**, за которым может следовать имя класса или экземпляр возбуждаемого исключения:\n",
    "\n",
    "        raise <instance>     # Возбуждает экземпляр класса-исключения\n",
    "        raise <class>        # Создает и возбуждает экземпляр класса-исключения\n",
    "        raise # Повторно возбуждает самое последнее исключение"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
